### PR TITLE
Parse config.ini for settings

### PR DESCRIPTION
**Pull request by thinkyhead**
- Add a `config.ini` file that can be parsed to modify configuration files before a build.
- Add a `CONFIG_DUMP` option to generate extra data files as part of the build.
- Add more `@section` markers to configuration files for easier navigation of the JSON.

---
`#define CONFIG_DUMP 2` to generate a `config.ini` file. Then do a build with any set of configurations. The `config.ini` file will be in the same folder as the `firmware.bin` / `firmware.hex` file.

Drop the populated `config.ini` into the `Marlin` folder (along with *any* Configuration files). Here are the base configs with almost every option disabled: [base_configs.tar.gz](https://github.com/MarlinFirmware/Marlin/files/9159908/base_configs.tar.gz) which provides a clean slate as the starting-point.

---
`#define CONFIG_DUMP 3` or `4` to export the configurations as a best-guess schema, gathering comments, options, and option requirements into an array of dictionaries organized by configuration file and `@section`. Option 3 dumps the data in JSON format, while option 4 dumps it in YAML format, for possible ingestion by Jekyll.

For the schema "requires" fields, all conditions are gathered together from parent `#if` and sibling `#elif` blocks. The schema isn't intended to be interpreted according to enabled options, but only to gather the requirements together for each option. However, it should be possible to do a second pass to determine which options are enabled (using line order), and then applying interpreters for each operator. We just need to compose a parsing tree to handle `ENABLED`, `ANY`, and other macros, plus value comparators. This was done previously using regex replacement and `eval` in Javascript, but should be simpler using a prebuilt parsing library.

---

The scripts `buildroot/share/PlatformIO/scripts/schema.py` and `configuration.py` are used by the build, but can also be executed as a standalone scripts to create the schema files within the Marlin folder or to process `config.ini` files outside of a build. These scripts may be expanded to produce more kinds of output, such as organized YML files for Jekyll documentation, etc.

### CODE REVIEW

**Comment by thisiskeithb** [comment]
Oh man. I thought this was a runtime configuration option before reading the description.

You got me excited!

**Comment by thinkyhead** [comment]
> runtime configuration option before reading the description.

- Would require the built firmware to know all possible configuration strings (i.e., their hashes) and how to convert their value strings into data structures. e.g., using `sscanf()`.
- Current runtime configuration is done by replaying the output of `M503`. We should always have G-codes for the lightweight configuration interface, and any "runtime config ini" should be made to map 1-to-1 with G-codes.
- To extend runtime configuration in Marlin, a first step would be to methodically implement all config-related G-codes as defined by RepRapFirmware, starting with basic geometric codes that define things like bed size, the number of probe points for a high level `G29`, and other options that users most want to change at runtime.
- Alternatively, implement the general-purpose pieces of `G29`: A probe point collector and set of G-codes that command Marlin to use the collected data, such as "use it for bilinear mesh."

Other thoughts after sleep….

File: Marlin/config.ini [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+default_retract_acceleration             = 3000
+
+default_minimumfeedrate                  = 0.0
+default_mintravelfeedrate                = 0.0
+
+minimum_planner_speed                    = 0.05
+min_steps_per_segment                    = 6
+default_minsegmenttime                   = 20000
+
+[basic]
```
Comment (rhapsodyv) [inline Marlin/config.ini]: maybe, we need a prefix do avoid any pio conflict?... like `[marlin:basic]`

**Comment by ucirello** [comment]
If this get merged, I believe the next step will be to make a tutorial using it to bisect bugs. I literally finished learning how to use opt_set and sketched a small movie telling how to do it. Should I hold it and wait for this PR to be merged first?

**Comment by thinkyhead** [comment]
> If this get merged, I believe the next step will be to make a tutorial using it to bisect bugs. I literally finished learning how to use `opt_set` and sketched a small movie telling how to do it. Should I hold it and wait for this PR to be merged first?

Maybe both are worth covering, since there are sure to be some times when you want to modify configuration from a shell script. It will be interesting to see how this add-on helps with git bisect, since one of the tricky things is always to get your settings to match between versions.

As for migration possibilities, we just have to compare each published config version and compare to the one that follows, and put those migration rules into their own Python script. Some of those migration rules might be automatically picked out of the sanity check files by just looking at the options that differ from one version to the other.

The next thing I want to do with this is to scrape for the configuration structure and dependencies and put all of that information into JSON and YAML formats –describing the complete configuration schema– and then bring that into the Jekyll site with templates, and then organize it from there. I've just started to learn about Python containers, and I can see those making this scraping / structuring work a lot easier.

**Comment by thisiskeithb** [comment]
[BigTreeTech’s TFT firmware](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware) uses a runtime configuration file named `config.ini`, so it may be a good idea to use a different name here to prevent possible conflicts.

File: Marlin/src/lcd/extui/ftdi_eve_touch_ui/ftdi_eve_lib/scripts/svg2cpp.py [hunk 2]
```diff
@@ -112,10 +112,10 @@ def from_svg_view_box(self, svg):
     if s:
       m = re.search('viewBox="([0-9-.]+) ([0-9-.]+) ([0-9-.]+) ([0-9-.]+)"', svg)
       if m:
-        self.x_min    = float(m.group(1))
-        self.y_min    = float(m.group(2))
-        self.x_max    = float(m.group(3))
-        self.y_max    = float(m.group(4))
+        self.x_min    = float(m]1])
```
Comment (rhapsodyv) [inline Marlin/src/lcd/extui/ftdi_eve_touch_ui/ftdi_eve_lib/scripts/svg2cpp.py]: ]] or []?

**Pull request merged**